Add new [Fact] and [Theory] attributes to generalize ConditionalFact

roncain opened the pull request:
This PR adds WcfFact and WcfTheory attributes and discoverers to take the place of
ConditionalFact and ConditionalTheory.  The new discoverers have
been generalized to reflect onto the test method for any of the new
"skip" attributes (IssueAttribute and ConditionAttribute).

For design discussion on this see issue #1034

roncain commented:
Reviewers: this PR is the result of my prototype to combine ConditionalFact and ActiveIssue.  My main goal of this PR is to solicit feedback about the approach.  See #1034 for deeper discussion comments.  If approved (subject to feedback), it can be checked in without affecting existing code.  And then a 2nd PR can make a sweep to change existing tests to use the new attributes.

roncain commented on src/System.Private.ServiceModel/tests/Scenarios/Infrastructure/WcfFactTests.cs:
```diff
@@ -0,0 +1,96 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Reviewers: this test case was included mostly to illustrate usage and to allow me to test.  I will delete it from the PR before merging.

roncain commented:
@hongdai -- I don't know whether these proposed changes overlap with your prototype for ETW tracing.  Please take a look and we can discuss if we need to merge our efforts.

mconnew commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/ConditionalTestDetectors.cs:
```diff
@@ -117,7 +117,7 @@ public static bool IsServerDomainJoined()
         public static bool IsWindows()
         {
-            return RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
+            return Platform.MatchesCurrentPlatform(PlatformID.Windows);
```
I'm curious, what scenario required this change?

roncain replied:
Now that I've centralized other Platform detection into the Platform class, I wanted to be consistent in using it.  Nothing required this change other than code consistency, and I can revert if you prefer.

mconnew replied:
I'm fine with code consistency being a reason to change this. I was wondering if there was some subtlety to comparing things differently that caused a slightly different behavior.

mconnew commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/xunit/ConditionAttribute.cs:
```diff
@@ -0,0 +1,121 @@
+    //  [Condition(nameof(Root_Certificate_Installed), nameof(Client_Certificate_Installed))]
+
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
+    public sealed class ConditionAttribute : WcfSkippableAttribute
```
>sealed class ConditionAttribute [](start = 11, length = 31)

I noticed that ConditionalFactAttribute is sealed too. Is there some architectural reason why this needs to be the case?

roncain replied:
They don't need to be sealed -- they are just test infrastructure.  So I'll remove that.

mconnew commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/xunit/ConditionAttribute.cs:
```diff
@@ -0,0 +1,121 @@
+            // A null or empty list of conditions is treated as "no conditions",
+            // and the test cases will not be skipped.
+            // Example: [Condition()] or [Condition((string[]) null)]
+            int conditionCount = Conditions == null ? 0 : Conditions.Length;
```
>Conditions == null [](start = 33, length = 18)

Is there a case where Conditions could be null? The only constructor has a params as the only argument. Doesn't passing no parameters still cause a zero length array to be passed? I know this is nit-picking, so I'm not expecting you to change anything, but could this be converted to a Contract.Assert  to protect against someone adding a different constructor in the future?

roncain replied:
When I first updated ConditionalFact, Stephen Toub observed it was legal for the attribute declaration to pass in a null explicitly, so asked for the check.

mconnew commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/xunit/ConditionAttribute.cs:
```diff
@@ -0,0 +1,121 @@
+                if (symbols.Length == 2)
+                {
+                    conditionMemberName = symbols[1];
+                    ITypeInfo type = testMethod.TestClass.Class.Assembly.GetTypes(false).Where(t => t.Name.Contains(symbols[0])).FirstOrDefault();
```
>FirstOrDefault [](start = 129, length = 14)

SingleOrDefault might be better in case there are multiple matches. I think we would prefer a failure if there's ambiguity rather than it being non-deterministic.

mconnew commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/xunit/ConditionAttribute.cs:
```diff
@@ -0,0 +1,121 @@
+                string[] symbols = conditionMemberName.Split('.');
+                Type declaringType = testMethodDeclaringType;
+
+                if (symbols.Length == 2)
```
This looks like we don't expect a type to have it's namespace. I think it would be worth adding logic so that if symbols.Length > 2, then presume the full namespace qualified type has been specified and instead of doing a search for any type with the same name, try and get the namespace qualified type exactly.

mconnew commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/xunit/ConditionAttribute.cs:
```diff
@@ -0,0 +1,121 @@
+                if ((conditionMethodInfo = LookupConditionalMethod(declaringType, conditionMemberName)) == null)
+                {
+                    falseConditions.Add(String.Format("Condition \"{0}\" not found in type \"{1}\".",
+                                                      conditionMemberName, testMethodDeclaringType.Name));
```
>testMethodDeclaringType [](start = 75, length = 23)

Use declaringType.FullName instead so that we know what type the method couldn't be found in.

mconnew commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/xunit/ConditionAttribute.cs:
```diff
@@ -0,0 +1,121 @@
+                catch (Exception exc)
+                {
+                    falseConditions.Add(String.Format("Condition \"{0}\" threw exception \"{1}\".",
+                                                      conditionMemberName, exc.ToString()));
```
>exc.ToString() [](start = 75, length = 14)

This is going to output the entire stack trace as part of the string (which will include new lines). The string text suggests you might have intended to just output the exception type and/or it's message. Just wanted to check the behavior matched the intention.

mconnew commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/xunit/ConditionAttribute.cs:
```diff
@@ -0,0 +1,121 @@
+                return mi;
+
+            PropertyInfo pi = ti.GetDeclaredProperty(name);
+            if (pi != null && pi.PropertyType == typeof(bool) && pi.GetMethod != null && pi.GetMethod.IsStatic && pi.GetMethod.GetParameters().Length == 0)
```
>pi.GetMethod.GetParameters() [](start = 114, length = 28)

Is it possible for a PropertyInfo to return a method from GetMethod which takes any parameters? Also, PropertyInfo.GetMethod will return non-public methods too. Is this what we want? This might cause us problems once .Net native is using this.

roncain replied:
Let me investigate this apart from the main PR (just to get it checked in first).  I copied this from BuildTools.

mconnew commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/Platform.cs:
```diff
@@ -0,0 +1,71 @@
+        public static string GetPlatformName(PlatformID id)
+        {
+            return Enum.Format(typeof(PlatformID), id, "G");
+        }
```
Could this be an extension method for PlatformID? That way you could get rid of the CurrentName property and instead do the following.
```csharp
    Platform.Current.GetName();
```

roncain replied:
Good idea -- will do.

mconnew commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/xunit/IssueAttribute.cs:
```diff
@@ -0,0 +1,50 @@
+
+        public PlatformID PlatformID { get; private set; }
+
+        public IssueAttribute(int issue, PlatformID platformID)
```
Is there any benefit to having a repo name here? Or can we assume that we will always have a tracking bug in our repo so all Issue id's will be for dotnet/wcf?

roncain replied:
I like the idea -- will add it.

mconnew commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/xunit/PlatformAttribute.cs:
```diff
@@ -0,0 +1,49 @@
+namespace Infrastructure.Common
+{
+    // The [Platform] attribute can be applied to any test method that
+    // is marked with [WcfFact] or [WcfTheory].  It indicates the which
```
>the [](start = 62, length = 3)

Nit: errant 'the' in comment

mconnew commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/xunit/PlatformAttribute.cs:
```diff
@@ -0,0 +1,49 @@
+    // If this attribute is absent, it means the test can run on all platforms.
+    // Examples: 
+    //  [Platform(PlatformID.Windows)]
+    //  [Platform(PlatformID.AnyUnix, Exclude=PlatformID.OSX)]
```
How would we use this to specify only exclude OSX? It looks like it would be this:
```csharp
    [Platform(PlatformID.Any, Exclude=PlatformID.OSX)]
```
Can we add a default constructor which sets PlatformID to PlatformID.Any? That way you can specify only the exclude on the attribute which I think is going to be a common use case.

roncain replied:
The next commit will have [Framework] and [OS] to allow this granularity.

mconnew commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/xunit/SkippedTestCase.cs:
```diff
@@ -0,0 +1,53 @@
+
+namespace Infrastructure.Common
+{
+    /// <summary>Wraps another test case that should be skipped.</summary>
```
Maybe add comment that this is a copy of [https://github.com/dotnet/buildtools/blob/master/src/xunit.netcore.extensions/SkippedTestCase.cs](https://github.com/dotnet/buildtools/blob/master/src/xunit.netcore.extensions/SkippedTestCase.cs)

roncain replied:
I was going to but @hongdai and I decided that this class will be renamed and diverge from BuildTools to be more extensible.  Actually, I think SkippedTestCase in BuildTools was copied from an xunit sample anyway.

mconnew commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/xunit/WcfFactDiscoverer.cs:
```diff
@@ -0,0 +1,31 @@
+{
+    public class WcfFactDiscoverer : FactDiscoverer
+    {
+        private readonly IMessageSink _diagnosticMessageSink;
```
You've only added this because the base class has its field holding this private. As this class isn't sealed, should you make this a protected property with private set?

roncain replied:
This was copied from buildtools, but let me mull over refactoring slightly when we see merge in changes from @hongdai

mconnew commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/xunit/WcfFactDiscoverer.cs:
```diff
@@ -0,0 +1,31 @@
+
+namespace Infrastructure.Common
+{
+    public class WcfFactDiscoverer : FactDiscoverer
```
Maybe add comment referencing [https://github.com/dotnet/buildtools/blob/master/src/xunit.netcore.extensions/Discoverers/ConditionalFactDiscoverer.cs](https://github.com/dotnet/buildtools/blob/master/src/xunit.netcore.extensions/Discoverers/ConditionalFactDiscoverer.cs) as inspiration for this in case things get refactored, it will be easier to know where to look to see what changes we need to make.

roncain replied:
I think we've diverged past recognition at this point, and even ConditionAttribute is diverging from ConditionalFact's logic.

mconnew commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/xunit/WcfTestDiscoverer.cs:
```diff
@@ -0,0 +1,100 @@
+    // to determine whether the test should be skipped.
+    internal class WcfTestDiscoverer
+    {
+
```
Nit: blank line

mconnew commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/xunit/WcfTestDiscoverer.cs:
```diff
@@ -0,0 +1,100 @@
+
+                if (issueSkipList.Count > 0)
+                {
+                    string skippedReason = string.Format("Active issue(s): {0}", string.Join(",", issueSkipList));
```
The output from IssueAttribute.GetSkipReason clashes a little with this. The output would look something like this:  
Active issue(s): issue 123 for "AnyUnix" skipped on "Linux", issue 125 for "Any" skipped on "Linux"  
There's redundancy as every issue will be reporting the current platform. You are already special casing the attribute here and you have a reference to a strongly typed IssueAttribute object. Maybe have GetSkipReason output:
```csharp
"{0} for \"{1}\", Issue, Platform.GetPlatformName(PlatformID)
```
And have WcfTestDiscoverer output:
```csharp
"Active issue(s): {0} skipped on {1}", string.Join(", ", issueSkipList), Platform.CurrentName
```

roncain replied:
Good point.  I've restructured a bit so the skip reasons only include the details and let the outer Format add the part that you observed was redundant.  We can tune the message if we find it confusing in practice.

hongdai commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/xunit/SkippedTestCase.cs:
```diff
@@ -0,0 +1,53 @@
+
+        public void Deserialize(IXunitSerializationInfo info) { _testCase.Deserialize(info); }
+
+        public Task<RunSummary> RunAsync(
```
Could we make this class WCFTestCase for WCF specific actions? I'll need to add ETW information in the RunAsync. I believe this class can be used for non skipped tests as well, if we don't set SkipReason.

roncain replied:
I'd prefer not to deviate from using "Fact" or "Theory" in the name, because those are fundamental xunit concepts.  I'm open to renaming the prefix part (example: TestFact and TestTheory or WcfTestFact and WcfTheoryFact).  Thoughts?

roncain commented:
Reviewers: I have incorporated feedback and added FrameworkAttribute, OSAttribute, and their corresponding enum and support types.  I have also added the concept of Framework and OS to the IssueAttribute.  I may need to tune the enum matching for OSAttribute as I see the console output from the CI or VSO runs, but I think that can be handled in a separate PR.  Remember, these changes are only additive and will not affect current tests.  A separate PR will happen after this one to convert all [Fact] to [WcfFact] before it affects existing tests.

I am interested in feedback on the new changes.  If you have comments, please indicate whether they must be fixed before merging or whether they can be fixed post-merge.  I would like to merge soon to enable @hongdai to modify WcfFact for ETW tracing.

/cc: @mconnew @StephenBonikowsky @iamjasonp @hongdai

roncain commented:
Reviewers: I've added finer-grained values for OSID to allow distinguishing at a better granularity.  But I've found trying to extract this from RuntimeInformation alone is fragile and would require non-deterministic string matching.  Please review PR #1397 at the same time, because I will need to check that in first. It will permit me to determine the OSID from the runtime package used for tests.

/cc: @mconnew @iamjasonp @StephenBonikowsky @hongdai

roncain commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/OS.cs:
```diff
@@ -0,0 +1,142 @@
+            return Enum.Format(typeof(OSID), id, "G");
+        }
+
+        private static OSID DetectCurrentOS()
```
Reviewers: it is my intent to replace this string matching against the OS description with a lookup table based on $(TestNugetRuntimeId) in TestProperties.  I will need to merge #1397 before I can do that.  I prefer that we merge this current PR asap to unblock @hongdai and let me switch to using TestProperties post-merge.

iamjasonp commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/Framework.cs:
```diff
@@ -0,0 +1,94 @@
+                return FrameworkID.NetFramework;
+            }
+
+            if (frameworkName.IndexOf(NetCoreFrameworkName, StringComparison.Ordinal) >= 0)
```
NIT: Ordinal or OrdinalIgnoreCase?

roncain replied:
These names are supposed to be an exact match against literals buried in RuntimeInformation. I think I want detection to fail and tests to report the problem if they change.

roncain commented:
Reviewers: please wait until I update this PR before reviewing.  I am reducing its scope a bit.  I will send out a new comment when it is ready for review again.

/cc: @Zhenlan @mconnew @iamjasonp @hongdai @StephenBonikowsky

mconnew commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/Framework.cs:
```diff
@@ -0,0 +1,94 @@
+                {
+                    _currentFrameworkID = DetectCurrentFramework();
+                    Console.WriteLine(String.Format("Detected current framework as \"{0}\" from description \"{1}\"",
+                                     _currentFrameworkID.Name(), CurrentFrameworkDescription));
```
Debugging output should be removed

roncain replied:
Perhaps in a later checkin after we see some VSO runs.  Right now, look at the Console logs on the CI runs and you can see which OS and Framework have been detected.  This would be a great debugging aid if you find tests being run unexpectedly on certain OSes or Frameworks.

mconnew commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/Framework.cs:
```diff
@@ -0,0 +1,94 @@
+            // The "unknown" in the skip reasons will indicate investigation is required.
+            Console.WriteLine(String.Format(
+                                "FrameworkID cannot be detected from RuntimeInformation.FrameworkDescription \"{0}\".", 
+                                frameworkName));
```
Debugging output should be removed

roncain replied:
Same comment as above -- leaving in at least for now because it provides vital information in the console logs across all the platforms.

mconnew commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/FrameworkID.cs:
```diff
@@ -0,0 +1,19 @@
+
+namespace Infrastructure.Common
+{
+    [Flags]
```
Why is this a flag? A current framework can't be multiple frameworks.

roncain replied:
Both FrameworkID and OSID are flags because they are bit masks.  They are bit masks to allow combinations like Any.  And you can combine them in attribute declarations to include or exclude certain frameworks like this:

```
    [WcfFact]
    [Issue(999, Framework = FrameworkID.Any ^ FrameworkID.NetFramework)]
    public static void Skip_Issue_All_Frameworks_Except_Full_Framework()
    {
    }

    [WcfFact]
    [Issue(999, Framework = FrameworkID.NetCore | FrameworkID.NetFramework)]
    public static void Skip_Issue_On_NetCore_And_Full_Framework()
    {
    }
```

mconnew commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/OS.cs:
```diff
@@ -0,0 +1,142 @@
+                if (_currentOSID == 0)
+                {
+                    _currentOSID = DetectCurrentOS();
+                    Console.WriteLine(String.Format("Detected current OS as \"{0}\" from description \"{1}\"",
```
Are these outputs intentional? I Don't think we should have them.

roncain replied:
Yes, these outputs are intentional, in the same way the Console writes are in other areas like certificate loading.  It captures information in the console logs on lab runs that help debug.  I'd like to leave these in at least for awhile and maybe permanently.

mconnew commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/OsID.cs:
```diff
@@ -0,0 +1,49 @@
+        None = 0,
+
+        Windows_7 = 1,
+        Windows_8_1 = 2,
```
Can we use hex values?

roncain replied:
ok

roncain commented:
Reviewers: this is ready to be reviewed now.  It has been reduced in scope.  As always, please indicate which comments are blocking and which can be addressed post-merge.  Thanks.

/cc: @Zhenlan @mconnew @iamjasonp @hongdai @StephenBonikowsky

mconnew commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/OsID.cs:
```diff
@@ -0,0 +1,49 @@
+        Ubuntu_16_04 = 4096,
+        AnyUbuntu = Ubuntu_14_04 | Ubuntu_16_04,
+
+        Unknown = 8192,     // were unable to determine
```
>Unknown = 8192 [](start = 8, length = 14)

Unknown shouldn't be the next number in sequence. It's not possible to have no OS, so maybe Unknown should have a value of 0 so that Unknown is the opposite of Any.

roncain replied:
I thought about this some more and decided it was better to remove 'Unknown' entirely and just return 'None' if we can't detect.  Then I added infrastructure tests that will fail if they ever encounter a None for OSID or FrameworkID.  This means that if we start running on some new framework or OS other than the ones we know here we will see a test failure, and we'll know we need to update the detection logic.

roncain commented:
Reviewers: all feedback incorporated and awaiting final sign-off from someone.
I will squash all spurious commits before merging.

(Note: look at the console output for any OuterLoop and search for "detected current" to verify that we really are detecting the OS and Framework on which we are running.  Waiting to see more in VSO runs. We can't yet distinguish between Win7, Win81 and Win10 -- I'll make a new PR when I learn how.)

/cc: @Zhenlan @mconnew @iamjasonp @hongdai @StephenBonikowsky

iamjasonp commented on src/System.Private.ServiceModel/tests/Common/Infrastructure/OsID.cs:
```diff
@@ -0,0 +1,59 @@
+        AnyUbuntu = Ubuntu_14_04 | Ubuntu_16_04,
+
+        AnyUnix = AnyCentOS | AnyDebian | AnyFedora | AnyOpenSUSE | AnyOSX | AnyRHEL | AnyUbuntu,
+        AnyWindows = Windows_7 | Windows_8_1 | Windows_10 | WindowsPhone | Windows_Nano,
```
>AnyWindows = Windows_7 | Windows_8_1 | Windows_10 | WindowsPhone | Windows_Nano, [](start = 8, length = 80)

NIT: this probably belongs near the "Windows" entries on Line 26

iamjasonp commented:
We should definitely look deeper into how to distinguish the different Windowses (sp? :smile:) 

LGTM

roncain commented:
@mconnew had an interesting suggestion how to refactor slightly to make IssueAttribute and ConditionAttribute both simpler and more powerful.  I am going to merge this PR as-is and open another one with his proposed changes.
